### PR TITLE
adds openshift image derivation test

### PR DESCRIPTION
**Pull request by cheesesashimi**
This PR introduces the OpenShift in-cluster image derivation test. What this does is introduce a new test which will test that in-cluster image derivation works as we intend. The test works thusly:

1. We grab the latest RHCOS image which was built just prior to test execution.
2. Build a derived image from the aforementioned RHCOS image.
3. Select a cluster node to apply the image to.
4. Apply the new image using `rpm-ostree`.
5. Reboot the cluster node.
6. Verify that the cluster node came up successfully.
7. Roll back applying the derived image.

This is a deceptively large PR because it introduces a bunch of vendored dependencies. For ease of review, I added the vendored dependencies in a commit by themselves. Additionally, since I wasn't sure about the impact my changes would have upon the `coreos-assembler` image, I opted to multistage the Dockerfile build so I could inject a pre-compiled version of the layering test suite into the `coreos-assembler` image.

The test itself originally came from https://github.com/openshift/machine-config-operator/pull/2886.

### CODE REVIEW

File: tests/layering/layering_test.go
Comment (mkenigs) [inline tests/layering/layering_test.go:118]:
```suggestion
	// rpm-ostree rebase --experimental ostree-unverified-image:image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/test-boot-in-cluster-image
```
or maybe just
```suggestion
	// rpm-ostree rebase --experimental imageURL
```
to avoid similar inconsistencies in the future

**Comment by jlebon** [comment]
Cool!

> The test itself originally came from https://github.com/openshift/machine-config-operator/pull/2886.

Is the test exactly the same? Could we instead of adding go vendoring here, have the builder also check out the MCO repo and build the test from there?

File: tests/layering/layering_test.go
Comment (cheesesashimi) [reply tests/layering/layering_test.go:118]: No worries! I'm cleaning this up as I go.

**Comment by cgwalters** [comment]
So...this is definitely tricky.  As of right now, this repository is only consumed by coreos-assembler - which already has a lot of Go code for tests.  Adding new Go code here means there's another place to handle vendor bumps etc.

Another balance we try to handle though is that we don't have a hard dependency for "coreos stuff" on OpenShift or even Kubernetes.  (e.g. coreos-assembler can also be run via podman)

WDYT about having this in e.g. https://github.com/openshift/origin/tree/master/test/extended instead?  I think we can wire things up so that the new test is only run on this repo to start - but it would mean we could also more easily run it as a periodic *outside* of PRs to this repo?

That said, on the other hand there *are* advantages to having tests live with the repository.  A big part of the reason [coreos external tests](https://github.com/coreos/coreos-assembler/blob/main/docs/kola/external-tests.md) were created is that the tests can live inside the config repo (like this one) that also (for FCOS) also versions the content.

I think I lean on balance towards adding this to the openshift tests, but it's by no means a strongly held opinion.  Another counterbalance could be adding *more* tests in Go code here.

**Comment by cgwalters** [comment]
Or I guess to say this another way what has me hesitating a bit is that today the output from a build of this repository is a "cosa build", which is the oscontainer image and disk images.

Now we *do* have a Dockerfile, but that's mainly to "fake out" Prow and try to get "a build" of stuff together - we don't actually ship the result of that build.

But if we're building a test, I think there are definitely cases where we'd want to build that test *outside* of building the repo - which is how the origin tests work.

If we were to go this path, perhaps actually it should be [fedora-coreos-config](https://github.com/coreos/fedora-coreos-config) that has a Go testing framework, and this adds RHCOS tests to it?  That would be a new middle ground between "external tests are in shell script" and "cosa tests which can be Go, but are lifecycled with cosa, not the OS".


Anyone else have opinions?

**Comment by jlebon** [comment]
Personally, I don't really mind Go code being used for tests in this repo or in f-c-c. I think  we've discussed in the past re. how if we want to keep draining kola of tests into the config repos, we'd probably need "cosa build"-level testing there to make `kola testiso` work in that model, and I can see Go being good for this for sharing with cosa.

I really dislike the Go vendoring model though. Maybe if we go that way, we can find some way to use the vendoring from cosa and not ship any vendoring in the config repos (which might be what you're implying already with "lifecycled with cosa").

But re. this test in specific, since this is a cluster-level test (which unlike kola tests, isn't really something developers are expected to easily run locally, right?) and OpenShift CI falls over anyway if GitHub is down, could we just ship `go.sum` and `go.mod`? I think that'd make this much easier to swallow.

**Comment by cheesesashimi** [comment]
I'm torn between this belonging here or in openshift/origin since this is a cluster-level test that tests interaction with RHCOS and various parts of OpenShift. I see benefits and drawbacks for both of those repos and I'm not particularly opinionated with it going in either of those places.

However, I'm not sure if it belongs with coreos-assembler or even fedora-coreos-config since those seem like they're more OS-specific tests that exercise and validate the various aspects of CoreOS. But I do see the value of vendoring the dependencies for this test within either of those repos. That said, I share @jlebon's dislike of the Go vendor model and think we can get away with only the `go.mod` / `go.sum` files in this repo.

**Comment by jlebon** [comment]
Also good with this as is.
/approve

Minor suggestion to layout: add a `tests/go` subdir and in there have the `go.mod` and `go.sum` files and the `layering/` subdir.

**Comment by cheesesashimi** [comment]
This is now ready for a more in-depth review.

File: tests/layering/layering_test.go
Comment (kikisdeliveryservice) [inline tests/layering/layering_test.go:30]: q: why %s here?

File: tests/layering/node_cmd_test.go
Comment (kikisdeliveryservice) [inline tests/layering/node_cmd_test.go:425]: q: probably obvious but why a pointer?

File: tests/layering/node_cmd_test.go
Comment (kikisdeliveryservice) [inline tests/layering/node_cmd_test.go:396]: q: if containers are ready and condition.status == true can the below ever be valid checks? (I think I'm asking: is containersReady a proxy for the check below??)

File: tests/layering/layering_test.go
Comment (kikisdeliveryservice) [inline tests/layering/layering_test.go:19]: nit: could we alphabetize these so they are grouped together a little more

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:135]:
```suggestion
	t.Log("placing image pull secret on node")
```

File: tests/layering/node_cmd_test.go
Comment (cheesesashimi) [reply tests/layering/node_cmd_test.go:396]: Honestly, I wasn't sure what I needed to check and went a bit overboard. Just this check alone might be sufficient.

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:161]: logf->log or alternatively add "node: %s, node.name" (or something)

File: tests/layering/node_cmd_test.go
Comment (cheesesashimi) [reply tests/layering/node_cmd_test.go:425]: The `Started` field is a pointer and one has to deference it in order to check if its true or false. I probably could've done something like:

```golang
if status.Started != nil && *status.Started != true {
  return false
}
```

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:224]: :)

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:312]:
```suggestion
	updatedNode, err := cs.Nodes().Get(context.TODO(), node.Name, metav1.GetOptions{})
```
You shouldn't need ObectMeta to get name for node of corev1.Node type for example in our repo: https://github.com/openshift/machine-config-operator/blob/9c6c2bfd7ed498bfbc296d530d1839bd6a177b0b/pkg/daemon/node.go#L24

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:325]: same as above re:objectmeta

File: tests/layering/layering_test.go
Comment (cheesesashimi) [reply tests/layering/layering_test.go:30]: This value gets replaced in with either `latest` or the OCP release (e.g., `4.11`) using https://github.com/openshift/os/pull/746/files#diff-58ffaa9005e1e047a4897d0a0d8a414238a5334c730449caa62fbd8fca7cb34dR36.

This occurs here: https://github.com/openshift/os/pull/746/files#diff-58ffaa9005e1e047a4897d0a0d8a414238a5334c730449caa62fbd8fca7cb34dR75.

My reasoning was that this could run on all of the release branches as well as `master`, so it should pull the appropriate image for that branch.

File: tests/layering/node_cmd_test.go
Comment (kikisdeliveryservice) [reply tests/layering/node_cmd_test.go:425]: i was thinking the same thing on that one :)

File: tests/layering/node_cmd_test.go
Comment (cheesesashimi) [inline tests/layering/node_cmd_test.go:1]: I wanted to mention that I may want to eventually port this library to the MCO's testing library. With this present, we could reduce our need to shell out to `oc` to execute arbitrary commands on a node.

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:161]: I made that mistake a lot; sorry! I'll get them all fixed and make a note to tune my linter.

File: tests/layering/node_cmd_test.go
Comment (kikisdeliveryservice) [reply tests/layering/node_cmd_test.go:1]: good idea, along those lines, can you add an owners file to this tests/layering?

File: tests/layering/layering_test.go
Comment (kikisdeliveryservice) [reply tests/layering/layering_test.go:30]: for some reason link not working what line is it (just for my own info when you get a chance)

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [inline tests/layering/helpers_test.go:361]: do we need a space or colon or something here ie ...+ " " + string(...)

Comment (cheesesashimi) [reply tests/layering/helpers_test.go:361]: I don't think we do, but it couldn't hurt!

File: tests/layering/layering_test.go
Comment (mkenigs) [inline tests/layering/layering_test.go:43]: We really need to make this a library somewhere :flushed: 
https://github.com/jkyros/machine-config-operator/pull/1#discussion_r819106902

File: tests/layering/helpers_test.go
Comment (mkenigs) [inline tests/layering/helpers_test.go:63]: Not sure how much it matters but I've been asked before to do `ctx := context.TODO()` and then use `ctx` when I'm calling `context.TODO()` repeatedly. You have that here and a few other places

File: tests/layering/helpers_test.go
Comment (mkenigs) [inline tests/layering/helpers_test.go:55]: Might be my imagination but it feels like a slightly more common pattern to call `Get` first and then `Create` second if it doesn't exist

File: tests/layering/layering_test.go
Comment (mkenigs) [reply tests/layering/layering_test.go:30]: tests/layering/helpers_test.go:75

File: tests/layering/helpers_test.go
Comment (mkenigs) [inline tests/layering/helpers_test.go:75]: How come you're grabbing this from an env var?

File: tests/layering/helpers_test.go
Comment (mkenigs) [inline tests/layering/helpers_test.go:125]: Does the `Create` call already return the build?

File: tests/layering/helpers_test.go
Comment (mkenigs) [inline tests/layering/helpers_test.go:160]:
```suggestion
		// Allow cleaning up the pull secret from the node if the node will not be rebooted or deleted
```

File: tests/layering/layering_test.go
Comment (mkenigs) [inline tests/layering/layering_test.go:103]: I got confused about whether it was the pull secret or the node, maybe
```suggestion
					// not reboot or get deleted, we should clean it up.
```

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:63]: What you suggested is good practice and something I should have been doing. Will fix.

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:55]: I could see that being the case. If that's what's more common, I'll adopt that pattern instead.

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:125]: Sort of. I don't think it's a complete build object. In other words, if you didn't supply some metadata, what it returns may be missing that metadata. However, I think that may be fine for what I'm doing with it.

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:312]: Yeah, I'll go through and fix all of those. I'm not sure why I did that ¯\_(ツ)_/¯

File: tests/layering/helpers_test.go
Comment (cheesesashimi) [reply tests/layering/helpers_test.go:361]: I have a slightly better idea to do that, somewhat inspired by how `os/exec` does it.

File: tests/layering/helpers_test.go
Comment (mkenigs) [reply tests/layering/helpers_test.go:55]: I dunno if it actually is though, I bet @kikisdeliveryservice would know?

File: tests/layering/layering_test.go
Comment (mkenigs) [reply tests/layering/layering_test.go:43]: We also have a duplicated `getRPMOStreeStatus`

File: tests/layering/helpers_test.go
Comment (kikisdeliveryservice) [reply tests/layering/helpers_test.go]: I'm a little confused bc I can't find this line anymore - but it depends on what you're doing (which I can't tell bc I can't find the line)

**Comment by cheesesashimi** [comment]
After a bunch of refactoring, I think this is ready for a final review.

**Review by cgwalters — APPROVED** [review]
Oops I started a review but forgot to click submit.
Nothing blocking though!

**Comment by cgwalters** [comment]
Is it correct the state of things that this just adds the test, but nothing executes it today?

Are you thinking we'd run the test in the periodic?

Nothing stops us from having an opt-in
`/test derivation`
here too?

OK but actually, don't we need to wire this test up to "launch a cluster" bits?

**Comment by cgwalters** [comment]
From my PoV though I'm good to `/lgtm` and iterate further after merge.  Will leave open for a bit in case anyone else has comments.

**Comment by cheesesashimi** [comment]
> Is it correct the state of things that this just adds the test, but nothing executes it today?

Yup. This just adds the test for now.

> Are you thinking we'd run the test in the periodic?

That was my plan. I think it would also be good to run on PRs as you mention below this.

> OK but actually, don't we need to wire this test up to "launch a cluster" bits?

Correct. That's being worked on in https://github.com/openshift/release/pull/26912. The "magic" bit over there is the `workflow: ipi-gcp` line which causes the OpenShift CI system to provision a cluster.

**Comment by kikisdeliveryservice** [comment]
I can give this a final pass this afternoon

**Comment by kikisdeliveryservice** [comment]
This looks great, Zack. 

/lgtm

**Pull request merged**
